### PR TITLE
Log operations in the Editing module

### PR DESCRIPTION
**Pull request by duartegalvao**
This PR adds logging for operations that might be destructive (such as comment edits or deletions) in the Editing module.

### CODE REVIEW

File: indico/modules/events/editing/operations.py [hunk 1]
```diff
@@ -269,38 +302,44 @@ def create_new_tag(event, code, title, color, system=False):
...
-        tag.title = title
-    if color:
-        tag.color = color
+def update_tag(tag, updates):
+    changes = tag.populate_from_dict({k: v for k, v in updates.items() if v is not None})
     db.session.flush()
     logger.info('Tag %r updated by %r', tag, session.user)
+    log_fields = {'code': 'Code', 'title': 'Title', 'color': 'Color'}
+    tag.event.log(EventLogRealm.management, LogKind.change, 'Editing', f'Tag {tag.verbose_title} updated',
+                  session.user, data={'Changes': make_diff_log(changes, log_fields)})
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:311]: By adding `meta={'editing_tag_id': tag.id}` to all the tag-related log calls (or by creating a `log()` method in `EditingTag` similar to the ones we have in `Contribution` etc.) one could easily use the filter button in the log viewer to see all operations related to that tag. Not super important for tags but maybe nice to use nonetheless whenever you are operating on a specific object that has an ID anyway.

File: indico/modules/events/editing/operations.py [hunk 2]
```diff
@@ -269,38 +302,44 @@ def create_new_tag(event, code, title, color, system=False):
     return tag
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:302]: Adding tags should be logged as well

File: indico/modules/events/editing/models/editable.py [hunk 3]
```diff
@@ -117,6 +117,10 @@ def locator(self):
     def event(self):
         return self.contribution.event
 
+    @property
+    def verbose_title(self):
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/editable.py:121]: Not sure if that's the best name. How about `log_title` since it's not suitable outside logging (no i18n + quoting that would probably look strange elsewhere)

File: indico/modules/events/editing/operations.py [hunk 4]
```diff
@@ -269,38 +302,44 @@ def create_new_tag(event, code, title, color, system=False):
...
     logger.info('Tag %r deleted by %r', tag, session.user)
     db.session.delete(tag)
+    tag.event.log(EventLogRealm.management, LogKind.negative, 'Editing', f'Tag {tag.verbose_title} deleted',
+                  session.user)
 
 
 def create_new_file_type(event, editable_type, **data):
     file_type = EditingFileType(event=event, type=editable_type)
-    file_type.populate_from_dict(data, keys=FILE_TYPE_ATTRS)
+    file_type.populate_from_dict(data, keys=FILE_TYPE_ATTRS.keys())
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:323]: Is this change actually needed? The way `keys` is used, passing a dict should work just fine (`if keys and key not in keys`)

File: indico/modules/events/editing/operations.py [hunk 5]
```diff
@@ -269,38 +302,44 @@ def create_new_tag(event, code, title, color, system=False):
...
     db.session.flush()
     logger.info('File type %r for %s created by %r', file_type, editable_type.name, session.user)
+    event.log(EventLogRealm.management, LogKind.positive, 'Editing', f'File type {file_type.name} created',
+              session.user)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:327]: Having the `editing_file_type_id` in the meta data would be nice here as well

File: indico/modules/events/editing/operations_test.py [hunk 6]
```diff
@@ -41,6 +46,7 @@ def test_cannot_undo_review_old_rev(dummy_contribution, dummy_user):
 def test_can_undo_review(db, dummy_contribution, dummy_user, type, ok):
     from indico.modules.events.editing.operations import InvalidEditableState, undo_review
     editable = Editable(contribution=dummy_contribution, type=EditableType.paper)
+    EditingRevision(editable=editable, user=dummy_user, type=RevisionType.ready_for_review)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations_test.py:49]: Why is this extra revision needed?

Comment (duartegalvao) [reply indico/modules/events/editing/operations_test.py:49]: Because I'm logging the editable state transition that happened when undoing the revision. If the editable has only one revision and it's undone, the new state will be null and and the logging fails.
Besides, it should never be possible to have an editable with zero valid revisions. Should I add extra checks/tests to ensure that? (since the first revision is always `new` or `ready_for_review`, it's not possible to undo them anyway, but could help make the logic clearer)

File: indico/modules/events/editing/models/tags.py [hunk 7]
```diff
@@ -66,6 +66,10 @@ def verbose_title(self):
         """Properly formatted title, including tag code."""
         return f'{self.code}: {self.title}'
 
+    def log(self, *args, **kwargs):
+        """Log with prefilled metadata for the tag."""
+        return self.event.log(*args, meta={'tag_id': self.id}, **kwargs)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/models/tags.py:71]:
```suggestion
        return self.event.log(*args, meta={'editing_tag_id': self.id}, **kwargs)
```

We also have registration tags and if we were to log anything for those, they should not "collide" (show up both) when filtering the event logs by metadata.

File: indico/modules/events/editing/operations.py [hunk 8]
```diff
@@ -260,47 +308,56 @@ def delete_revision_comment(comment):
...
     logger.info('Tag %r updated by %r', tag, session.user)
+    log_fields = {'code': 'Code', 'title': 'Title', 'color': 'Color'}
+    tag.log(EventLogRealm.management, LogKind.change, 'Editing', f'Tag {tag.verbose_title} updated', session.user,
+            data={'Changes': make_diff_log(changes, log_fields)})
 
 
 def delete_tag(tag):
     logger.info('Tag %r deleted by %r', tag, session.user)
     db.session.delete(tag)
+    tag.log(EventLogRealm.management, LogKind.negative, 'Editing', f'Tag {tag.verbose_title} deleted', session.user)
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:336]: Probably safer to do it the other way round. SA shouldn't be clearing anything in the deleted model, but it feels a bit dirty to call something on a deleted object...

```suggestion
    tag.log(EventLogRealm.management, LogKind.negative, 'Editing', f'Tag {tag.verbose_title} deleted', session.user)
    db.session.delete(tag)
```

File: indico/modules/events/editing/operations.py [hunk 9]
```diff
@@ -260,47 +308,56 @@ def delete_revision_comment(comment):
...
     db.session.flush()
     logger.info('File type %r for %s created by %r', file_type, editable_type.name, session.user)
+    file_type.log(EventLogRealm.management, LogKind.positive, 'Editing', f'File type {file_type.name} created',
+                  session.user)
     return file_type
 
 
 def update_file_type(file_type, **data):
-    file_type.populate_from_dict(data, keys=FILE_TYPE_ATTRS)
+    changes = file_type.populate_from_dict(data, keys=FILE_TYPE_ATTRS.keys())
```
Comment (ThiefMaster) [inline indico/modules/events/editing/operations.py:350]:
```suggestion
    changes = file_type.populate_from_dict(data, keys=FILE_TYPE_ATTRS)
```

untested but I think this isn't needed

**Review by ThiefMaster — APPROVED** [review]
